src/optimize_qdq_format.cpp

```diff
@@ -0,0 +1,169 @@
+    {
+        return match::name("dequantizelinear")(
+            match::arg(0)(match::name("quantizelinear")(
+                              match::arg(0)(match::name(get_quantizable_op_names()).bind("qop")))
+                              .bind("quantizelinear")));
+    }
+
+    void apply(module& m, match::matcher_result r) const
+    {
+        auto dq       = r.result;
+        auto ins      = r.instructions["qop"];
+        auto args     = ins->inputs();
+        bool replace  = false;
+        bool mismatch = false;
+
+        for(auto&& arg : args)
+        {
+            if(arg->name() == "dequantizelinear")
+            {
+                auto q1 = arg->inputs().front();
+                if(q1->get_shape().type() == migraphx::shape::int8_type)
+                {
+                    arg     = q1;
+                    replace = true;
+                }
+                else
+                {
+                    mismatch = true;
+                }
+            }
+        }
+
+        if(replace and not mismatch)
+        {
+            if(ins->name() == "convolution")
+            {
+                auto conv_op = any_cast<op::convolution>(ins->get_operator());
+                m.replace_instruction(ins,
+                                      op::quant_convolution{conv_op.padding,
+                                                            conv_op.stride,
+                                                            conv_op.dilation,
+                                                            conv_op.padding_mode,
+                                                            conv_op.group},
+                                      args);
+            }
+            else if(ins->name() == "dot")
+            {
+                auto dot_op = any_cast<op::dot>(ins->get_operator());
+                m.replace_instruction(ins,
+                                      op::quant_dot{static_cast<int32_t>(dot_op.alpha),
+                                                    static_cast<int32_t>(dot_op.beta)},
+                                      args);
+            }
+            else
+            {
+                MIGRAPHX_THROW(ins->name() +
+                               " does not currently have a quantized implementation.");
+            }
+            auto dq_args    = dq->inputs();
+            dq_args.front() = ins;
+            m.replace_instruction(dq, dq->get_operator(), dq_args);
+        }
+    }
+};
+
+void remove_qdq_pairs(module& m)
+{
+    for(auto ins : iterator_for(m))
+    {
+        bool replace_op = false;
+        auto args       = ins->inputs();
+        for(auto&& arg : args)
+        {
+            if(arg->name() == "dequantizelinear")
+            {
+                auto q = arg->inputs().front();
+                if(q->name() == "quantizelinear")
+                {
+                    arg        = q->inputs().front();
+                    replace_op = true;
+                }
+            }
+        }
```

Comment:
Use `std::find_if` here instead of the for loop.

Reply:
I think we need to be able to iterate over all inputs in case multiple input tensors have QDQs. Using find_if would stop after the first input with QDQ.